src/main/java/com/linecorp/armeria/server/metrics/MetricsServiceCodec.java

```diff
@@ -0,0 +1,71 @@
+package com.linecorp.armeria.server.metrics;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+
+import com.linecorp.armeria.common.ServiceInvocationContext;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.metrics.RequestMetrics;
+import com.linecorp.armeria.server.DecoratingServiceCodec;
+import com.linecorp.armeria.server.ServiceCodec;
+import com.linecorp.armeria.server.ServiceConfig;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.util.concurrent.Promise;
+
+class MetricsServiceCodec extends DecoratingServiceCodec {
+
+    private final String serviceName;
+    private final MetricRegistry metricRegistry;
+    private final Map<String, RequestMetrics> methodRequestMetrics;
+
+    /**
+     * Creates a new instance that decorates the specified {@link ServiceCodec}.
+     */
+    MetricsServiceCodec(ServiceCodec delegate, String serviceName, MetricRegistry metricRegistry) {
+        super(delegate);
+        this.serviceName = serviceName;
+        this.metricRegistry = metricRegistry;
+        methodRequestMetrics = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public DecodeResult decodeRequest(ServiceConfig cfg, Channel ch, SessionProtocol sessionProtocol,
+                                      String hostname, String path, String mappedPath, ByteBuf in,
+                                      Object originalRequest, Promise<Object> promise) throws Exception {
+        DecodeResult decodeResult = delegate().decodeRequest(
+                cfg, ch, sessionProtocol, hostname, path, mappedPath, in, originalRequest, promise);
+        if (decodeResult.type() == DecodeResultType.SUCCESS) {
+            ServiceInvocationContext ctx = decodeResult.invocationContext();
+            String methodLoggedName = RequestMetrics.metricName(ctx);
+            RequestMetrics metrics = getRequestMetrics(
+                    MetricRegistry.name("server", serviceName, ctx.method()));
+            Timer.Context timer = metrics.time();
+            promise.addListener(future -> {
+                timer.stop();
+                if (future.isSuccess()) {
```

Comment:
don't we need to check its status code? i.e. is it successful even if future is `500` or so?